pkg/model/model.go

```diff
@@ -0,0 +1,90 @@
+package model
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SessionContext holds the context for a single session object, giving access to key things like REST Client and target Namespace
+type SessionContext struct {
+	context.Context
+
+	Name      string
+	Namespace string
+	Client    client.Client
+	Log       logr.Logger
+}
+
+// Ref references to a single Reference, e.g. Deployment, DeploymentConfig or GitRepo
+type Ref struct {
+	Name             string
+	ResourceStatuses []ResourceStatus
+}
+
+// AddResourceStatus adds the status of an involved Resource to this ref
+func (r *Ref) AddResourceStatus(ref ResourceStatus) {
+	replaced := false
+	for i, status := range r.ResourceStatuses {
+		if ref.Name == status.Name && ref.Kind == status.Kind {
+			r.ResourceStatuses[i] = ref
+			replaced = true
+		}
+	}
+	if !replaced {
+		r.ResourceStatuses = append(r.ResourceStatuses, ref)
+	}
+}
+
+// RemoveResourceStatus removes the status of an involved Resource to this ref
+func (r *Ref) RemoveResourceStatus(ref ResourceStatus) {
+	for i, status := range r.ResourceStatuses {
+		if ref.Name == status.Name && ref.Kind == status.Kind {
+			r.ResourceStatuses = append(r.ResourceStatuses[:i], r.ResourceStatuses[i+1:]...)
+		}
+	}
+}
+
+// GetResourceStatus returns a array of involved Resources based on a k8 Kind
+func (r *Ref) GetResourceStatus(kind string) []ResourceStatus {
+	refs := []ResourceStatus{}
+	for _, status := range r.ResourceStatuses {
+		if status.Kind == kind {
+			refs = append(refs, status)
+		}
+	}
+	return refs
+}
+
+// ResourceStatus holds information about the resources created/changed to fulfill a Ref
+type ResourceStatus struct {
+	Kind string
+	Name string
+	// Created, Mutated
+	Action ResourceAction
+}
+
+// ResourceAction describes which type of operation was done/attempted to the target resource. Used to determine how to undo it.
+type ResourceAction string
+
+const (
+	// ActionCreated imply the whole Named Kind was created and can be deleted
+	ActionCreated ResourceAction = "created"
+	// ActionModified imply the Named Kind has been modified and needs to be reverted to get back to original state
+	ActionModified ResourceAction = "modified"
+	// ActionFailed imply what ever was attempted failed. Assume current state is ok in clean up?
+	ActionFailed ResourceAction = "failed"
+)
+
+// Locator should attempt to resolve a Ref.Name to a target Named kind, e.g. Deployment, DeploymentConfig.
+// return false if nothing was found
+type Locator func(SessionContext, *Ref) bool
```

Comment:
Wondering if we shouldn't return found instance instead of just `bool` - in general what is the idea behind that entity on its own?

Reply:
Yeah, my thinking here was that it would need to return some type of 'Target' object, and that 'Target' object is then passed on to the Mutator/Revertors. 

e.g. 

``go
type Target struct {
   Kind string
   Name string
   Labels map[string]string
}
```

Reply:
https://github.com/aslakknutsen/istio-workspace/issues/71